File: apps/android/app/src/main/java/com/exponential/app/domain/IssueFilters.kt
package com.exponential.app.domain

// Mirrored from apps/web/src/lib/filters.ts. The filter shape and matching
// semantics must match the web mapping; if you change one, change the others
// (apps/web/src/lib/filters.ts, apps/ios/.../Domain/IssueFilters.swift).
//
// EXP-314: the status filter is a set of GROUP KEYS — a team status row id, or
// `builtin:<key>` for a constructed fallback the shape hasn't synced yet — not
// the anchor enum, exactly like labelIds.
data class IssueFilters(
    val statusIds: Set<String> = emptySet(),
    val priorities: Set<IssuePriority> = emptySet(),
    val labelIds: Set<String> = emptySet(),
) {
    val isEmpty: Boolean get() =
        statusIds.isEmpty() && priorities.isEmpty() && labelIds.isEmpty()
    val count: Int get() =
        statusIds.size + priorities.size + labelIds.size
}

fun matchesFilters(
    statusGroupKey: String,
    priority: IssuePriority,
    issueLabelIds: Collection<String>,
    filters: IssueFilters,
): Boolean {
    if (filters.statusIds.isNotEmpty() && statusGroupKey !in filters.statusIds) return false
    if (filters.priorities.isNotEmpty() && priority !in filters.priorities) return false
    if (filters.labelIds.isNotEmpty() && filters.labelIds.none { it in issueLabelIds }) return false
    return true
}

File: apps/ios/ExpCore/Sources/Domain/IssueFilters.swift
import Foundation

// Mirrored from apps/web/src/lib/filters.ts. The filter shape and matching
// semantics must match the web mapping; if you change one, change the others
// (apps/web/src/lib/filters.ts, apps/android/.../domain/IssueFilters.kt).
public struct IssueFilters: Equatable, Sendable {
    /// EXP-314: status GROUP KEYS — an `issue_statuses` row id, or the
    /// synthetic `builtin:<key>` of a constructed default while the statuses
    /// shape hasn't synced. Same semantics as `labelIds`.
    public var statusIds: Set<String> = []
    public var priorities: Set<IssuePriority> = []
    public var labelIds: Set<String> = []

    public init(statusIds: Set<String> = [], priorities: Set<IssuePriority> = [], labelIds: Set<String> = []) {
        self.statusIds = statusIds
        self.priorities = priorities
        self.labelIds = labelIds
    }

    public var isEmpty: Bool { statusIds.isEmpty && priorities.isEmpty && labelIds.isEmpty }
    public var count: Int { statusIds.count + priorities.count + labelIds.count }
}

/// `statusId` is the issue's RESOLVED group key (`ResolvedIssueStatus.id`), not
/// the raw `issues.status_id` column — a pre-backfill row still matches the
/// builtin group it resolves into.
public func matchesFilters(
    statusId: String,
    priority: IssuePriority,
    issueLabelIds: Set<String>,
    filters: IssueFilters
) -> Bool {
    if !filters.statusIds.isEmpty && !filters.statusIds.contains(statusId) { return false }
    if !filters.priorities.isEmpty && !filters.priorities.contains(priority) { return false }
    if !filters.labelIds.isEmpty && filters.labelIds.isDisjoint(with: issueLabelIds) { return false }
    return true
}
